subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CleanupOutputsStep.java

```diff
@@ -0,0 +1,92 @@
+
+import com.google.common.collect.ImmutableSet;
+import org.gradle.internal.execution.IncrementalContext;
+import org.gradle.internal.execution.InputChangesContext;
+import org.gradle.internal.execution.Result;
+import org.gradle.internal.execution.Step;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.impl.OutputsCleaner;
+import org.gradle.internal.file.TreeType;
+import org.gradle.internal.fingerprint.FileCollectionFingerprint;
+import org.gradle.util.GFileUtils;
+
+import java.io.File;
+
+public class CleanupOutputsStep implements Step<InputChangesContext, Result> {
+
+    private final Step<? super InputChangesContext, ? extends Result> delegate;
+
+    public CleanupOutputsStep(Step<? super InputChangesContext, ? extends Result> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Result execute(InputChangesContext context) {
+        if (!context.isIncrementalExecution()) {
+            UnitOfWork work = context.getWork();
+            if (work.isCleanupOutputsOnNonIncrementalExecution()) {
+                if (work.isOverlappingOutputsDetected()) {
+                    cleanupOverlappingOutputs(context, work);
+                } else {
+                    cleanupExclusiveOutputs(work);
+                }
+            }
+        }
+        return delegate.execute(context);
+    }
+
+    private void cleanupOverlappingOutputs(IncrementalContext context, UnitOfWork work) {
+        context.getAfterPreviousExecutionState().ifPresent(previousOutputs -> {
+            ImmutableSet.Builder<File> builder = ImmutableSet.builder();
+            work.visitOutputProperties((name, type, roots) -> {
+                if (type == TreeType.DIRECTORY) {
+                    builder.addAll(roots);
+                }
+                if (type == TreeType.FILE) {
+                    for (File root : roots) {
+                        File parentFile = root.getParentFile();
+                        if (parentFile != null) {
+                            builder.add(parentFile);
+                        }
+                    }
+                }
+            });
+            ImmutableSet<File> preparedOutputDirectories = builder.build();
+            OutputsCleaner cleaner = new OutputsCleaner(file -> true, file -> !preparedOutputDirectories.contains(file));
+            for (FileCollectionFingerprint fileCollectionFingerprint : previousOutputs.getOutputFileProperties().values()) {
+                cleaner.cleanupOutputs(fileCollectionFingerprint);
+            }
+        });
+    }
+
+    private void cleanupExclusiveOutputs(UnitOfWork work) {
+        work.visitOutputProperties((name, type, roots) -> {
+            for (File root : roots) {
+                if (root.exists()) {
+                    if (type == TreeType.DIRECTORY) {
+                        GFileUtils.cleanDirectory(root);
+                    } else {
+                        GFileUtils.deleteFileQuietly(root);
```

Comment:
This doesn't ensure the file is deleted.

Reply:
I am using `forceDelete` now.